scripts/canaryRelease.mts

```diff
@@ -0,0 +1,242 @@
+    return;
+  }
+
+  await confirmPublishing(changedPackages, yes);
+
+  try {
+    await setAccessToken(accessToken);
+    await setVersion(changedPackages);
+    await buildPackages();
+    await publishPackages(changedPackages, dryRun);
+  } finally {
+    await cleanUp();
+  }
+}
+
+async function ensureCleanWorkingDirectory() {
+  try {
+    await $`git diff --quiet`;
+    await $`git diff --quiet --cached`;
+  } catch (error) {
+    console.error('❌ Working directory is not clean.');
+    process.exit(1);
+  }
+}
+
+async function listPublicChangedPackages(baseline: string) {
+  const { stdout: packagesJson } =
+    await $`pnpm list --recursive --filter ...[${baseline}] --depth -1 --only-projects --json`;
+  const packages = JSON.parse(packagesJson) as PackageInfo[];
+  return packages.filter((pkg) => !pkg.private);
+}
+
+async function getChangedPackages(
+  baseline: string | undefined,
+  skipLastCommitComparison: boolean,
+): Promise<PackageInfo[]> {
+  if (!skipLastCommitComparison) {
+    const publicPackagesUpdatedInLastCommit = await listPublicChangedPackages('HEAD~1');
+    if (publicPackagesUpdatedInLastCommit.length === 0) {
+      console.log('No public packages changed in the last commit.');
+      return [];
+    }
+  }
+
+  if (!baseline) {
+    const { stdout: latestTag } = await $`git describe --abbrev=0`;
+    baseline = latestTag;
+  }
+
+  console.log(`Looking for changed public packages since ${chalk.yellow(baseline)}...`);
+
+  const changedPackages = await listPublicChangedPackages(baseline);
+  if (changedPackages.length === 0) {
+    console.log('Nothing found.');
+  }
+
+  return changedPackages;
+}
+
+async function confirmPublishing(changedPackages: PackageInfo[], yes: boolean) {
+  if (!yes) {
+    const rl = readline.createInterface({
+      input: process.stdin,
+      output: process.stdout,
+    });
+
+    console.log('\nFound changes in the following packages:');
+    for (const pkg of changedPackages) {
+      console.log(` - ${pkg.name}`);
+    }
+
+    console.log('\nThis will publish the above packages to the npm registry.');
+    const answer = await rl.question('Do you want to proceed? (y/n) ');
+
+    rl.close();
+
+    if (answer.toLowerCase() !== 'y') {
+      console.log('Aborted.');
+      process.exit(0);
+    }
+  }
+}
+
+async function setAccessToken(npmAccessToken: string | undefined) {
+  if (!npmAccessToken && !process.env.NPM_TOKEN) {
+    console.error(
+      '❌ NPM access token is required. Either pass it as an argument or set it as an NPM_TOKEN environment variable.',
+    );
+    process.exit(1);
+  }
+
+  const npmrcPath = resolve(workspaceRoot, '.npmrc');
+
+  await appendFile(
+    npmrcPath,
+    `//registry.npmjs.org/:_authToken=${npmAccessToken ?? process.env.NPM_TOKEN}\n`,
+  );
+}
+
+async function setVersion(packages: PackageInfo[]) {
+  const { stdout: currentRevisionSha } = await $`git rev-parse --short HEAD`;
+  const timestamp = formatDate(new Date());
```

Comment:
🤔 I was just thinking, we should be able to read the date from the commit. That way we make it impossible to publish the same commit twice.

Reply:
What's the issue with publishing the same commit twice?

Reply:
To clarify, "Twice occasinally" is not really my main concern. It's about making this script idempotent. When multiple runs of the script has the same outcome. This makes it have an inherent resiliency which comes in handy when we're going to automate it. It becomes safe to do automatic retries, and an accidental misconfiguration can't make you end up with 100 times publishing the same package.

Reply:
Good point!